src/networkd.c

```diff
@@ -151,6 +152,72 @@ write_tunnel_params(GString* s, const NetplanNetDefinition* def)
+{
+    GString *params = NULL;
+    params = g_string_sized_new(200);
+
+    g_assert(def->tunnel.input_key);
+    /* The "PrivateKeyFile=" setting is available as of systemd-netwokrd v242+
+     * Base64 encoded PrivateKey= or absolute PrivateKeyFile= fields are mandatory.
+     *
+     * The key was already validated via validate_tunnel_grammar(), but we need
+     * to differentiate between base64 key VS absolute path key-file. And a base64
+     * string could (theoretically) start with '/', so we use is_wireguard_key()
+     * as well to check for more specific characteristics (if needed). */
+    if (def->tunnel.input_key[0] == '/' && !is_wireguard_key(def->tunnel.input_key))
+        g_string_append_printf(params, "PrivateKeyFile=%s\n", def->tunnel.input_key);
+    else
+        g_string_append_printf(params, "PrivateKey=%s\n", def->tunnel.input_key);
+
+    if (def->tunnel.port)
+        g_string_append_printf(params, "ListenPort=%u\n", def->tunnel.port);
+    /* This is called FirewallMark= as of systemd v243, but we keep calling it FwMark= for
+       backwards compatibility. FwMark= is still supported, but deprecated:
+       https://github.com/systemd/systemd/pull/12478 */
+    if (def->tunnel.fwmark)
+        g_string_append_printf(params, "FwMark=%u\n", def->tunnel.fwmark);
+
+    g_string_append_printf(s, "\n[WireGuard]\n%s", params->str);
+    g_string_free(params, TRUE);
+
+    for (guint i = 0; i < def->routes->len; i++) {
+        NetplanIPRoute *peer = g_array_index (def->routes, NetplanIPRoute*, i);
+        /* Routes which are not of "type: wireguard" are not for us */
+        if (!!g_ascii_strcasecmp(peer->type, "wireguard")) continue;
+
+        GString *peer_s = g_string_sized_new(200);
+
+        g_string_append_printf(peer_s, "PublicKey=%s\n", peer->public_key);
+        g_string_append(peer_s, "AllowedIPs=");
+        for (guint i = 0; i < peer->allowed_ips->len; ++i) {
+            if (i > 0 )
+                g_string_append_c(peer_s, ',');
+            g_string_append_printf(peer_s, "%s", g_array_index(peer->allowed_ips, char*, i));
+        }
+        g_string_append_c(peer_s, '\n');
+
+        if (peer->keepalive)
+            g_string_append_printf(peer_s, "PersistentKeepalive=%d\n", peer->keepalive);
+        if (peer->via)
+            g_string_append_printf(peer_s, "Endpoint=%s\n", peer->via);
+        /* The key was already validated via validate_tunnel_grammar(), but we need
+         * to differentiate between base64 key VS absolute path key-file. And a base64
+         * string could (theoretically) start with '/', so we use is_wireguard_key()
+         * as well to check for more specific characteristics (if needed). */
+        if (peer->preshared_key) {
+            if (peer->preshared_key[0] == '/' && !is_wireguard_key(peer->preshared_key))
+                g_string_append_printf(peer_s, "PresharedKeyFile=%s\n", peer->preshared_key);
```

Comment:
Is the `PresharedKeyFile` similarly only available in systemd 242+ ? Thinking about it now, maybe we should include some info in the documentation about when, with which dependency versions, the absolute keyfile paths can be used for it to work.

Reply:
+1 I updated the docs.